state/indexer/block/kv/util.go

```diff
@@ -51,6 +51,51 @@ func heightKey(height int64) ([]byte, error) {
 	)
 }
 
+func keyBelongsToHeightRange(key []byte, left, right int64) bool {
+	// left included, right excluded
+	eventHeight, err := parseHeightFromEventKey(key)
+	if err == nil {
+		return left <= eventHeight && eventHeight < right
+	}
+
+	var (
+		blockHeightKeyPrefix string
+		possibleHeight       int64
+	)
+	remaining, err := orderedcode.Parse(string(key), &blockHeightKeyPrefix, &possibleHeight)
+	if err != nil {
+		return false
+	}
+	if len(remaining) == 0 &&
+		blockHeightKeyPrefix == types.BlockHeightKey &&
+		left <= possibleHeight &&
+		possibleHeight < right {
+		return true
+	}
+	return false
+}
+
+func getHeightFromKey(key []byte) int64 {
+	// Must be called with either heightKey or eventKey
+	eventHeight, err := parseHeightFromEventKey(key)
+	if err == nil {
+		return eventHeight
+	}
+
+	var (
+		blockHeightKeyPrefix string
+		possibleHeight       int64
+	)
+	remaining, err := orderedcode.Parse(string(key), &blockHeightKeyPrefix, &possibleHeight)
+	if err != nil {
+		panic(err)
```

Comment:
Just for my own instruction, why do we panic here after calling `orderedcode.Parse`, but we just error out above (line 67)?

Reply:
[keyBelongsToHeightRange](https://github.com/cometbft/cometbft/blob/18269cff546a7a3b67fe27179c8c6168f21582f6/state/indexer/block/kv/util.go#L54C6-L54C29) function requires no guarantees on the key passed to it. Its goal is to check whether the given key belongs to a range of heights. That might not be the case for two reasons: the key belongs to the height out of the specified range, or the key doesn't belong to any height, meaning it's neither `heightKey` nor `eventKey`. In the latter case, we might fail to parse the key (the bulk of random bytes) into `heightKey`. That's nothing critical, but simply means that it is not the key we are interested in. 

Whereas, [getHeightFromKey](https://github.com/cometbft/cometbft/blob/18269cff546a7a3b67fe27179c8c6168f21582f6/state/indexer/block/kv/util.go#L78C6-L78C22) [requires](https://github.com/cometbft/cometbft/blob/18269cff546a7a3b67fe27179c8c6168f21582f6/state/indexer/block/kv/util.go#L79) its argument key to be the key with height (either `heightKey` or `eventKey`).  So in case it's not, we panic. 

I wrote comments for these to functions to make it more clear.